### PR TITLE
Call AsyncIteratorMethodBuilder.Complete before SetResult

### PR DESCRIPTION
**Pull request by jcouv**
While investigating a leak in async-iterator scenario, Stephen [pointed out](https://github.com/dotnet/runtime/issues/41187#issuecomment-678696730) that we should call `AsynciteratorMethodBuilder.Complete` sooner (before notifying caller of result or exception).
It doesn't looks like that's the cause of the leak, but is still worthwhile to fix.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Comment by jcouv** [comment]
FYI @stephentoub. Let me know if any second thoughts on this ordering. Thanks

**Comment by jcouv** [comment]
@cston @AlekseyTs @jaredpar for a second review. Thanks

**Comment by stephentoub** [comment]
@jcouv, I'm fairly certain the new ordering is the right ordering, and I'm a little surprised it didn't fix the cited issue.  Now that I'm back from being out, I'll find time to investigate the cited issue.  Feel free to hold off on this until then if you don't mind having the open PR around.

**Comment by jcouv** [comment]
@stephentoub Okay. I'll mark the PR as draft for now. Feel free to use this branch (open Roslyn.sln, hit F5) to try the fix. 
I was careful in testing impact on leak, so I'm fairly confident despite the variability in repro (I tried a few times without fix and a few times with).

**Comment by jcouv** [comment]
@stephentoub Did you get a chance to take a look at the leak? Should I move forward with this change? Let me know if I can help.

**Comment by stephentoub** [comment]
Thanks, Julien.  Looking at it now.

**Comment by stephentoub** [comment]
@jcouv, go ahead with this change.  I finally made time with a debugger and figured out the problem.  This PR is necessary but not sufficient.  Will describe the problem in the https://github.com/dotnet/runtime/issues/41187.

**Comment by jcouv** [comment]
@cston @AlekseyTs @jaredpar for a second review. Thanks

**Review by cston — APPROVED** [review]

**Pull request merged**
